Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
 				\param	pos
 				\param	polygon
 				\param	polygonIndices
-				\return conversionFailed true/false
 				\note	At the end, the calculated and merged face vertices are handed over to the \c triangulateFace function.
 				*/
 
-				bool convertIfcFace(const EXPRESSReference<typename IfcEntityTypesT::IfcFace>& face,
+				void convertIfcFace(const EXPRESSReference<typename IfcEntityTypesT::IfcFace>& face,
 					const carve::math::Matrix& pos,
-					std::shared_ptr<carve::input::PolyhedronData> polygon,
+					std::shared_ptr<carve::input::PolyhedronData>& polygon,
 					std::map<std::string, uint32_t>& polygonIndices)  const throw(...) 
 				{
 					if (face.expired()) {
 						throw oip::ReferenceExpiredException(face);
 					}
-					// Indicates if conversion has failed
-					bool conversionFailed = false;
 
-					// Id of face in step file
-					const uint32_t faceID = face->getId();
+					// get attribute 1:
+					// list of bound loops (IfcFaceBound) with outer boundary loop (IfcFaceOuterBound) at index 0;
+					// one bound loop contains a list of loop points (carve::gem::vector<3>)
+					std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops;
+					convertIfcFaceBoundList(face->Bounds, pos, faceBoundLoops);
 
-					// To triangulate the mesh, carve needs 2D polygons, we collect the data in 2D and 3D for every bound
-					std::vector<std::vector<carve::geom2d::P2>> faceVertices2D;
-					std::vector<std::vector<carve::geom::vector<3>>> faceVertices3D;
+					// if simple case: one triangle without inner bound (= hole)
+					//  -> simple direct triangle construction is possible (probably fast)
+					if ((faceBoundLoops.size() == 1) && (faceBoundLoops[0].size() == 3))
+					{
+						// std::vector<carve::geom::vector<3>> loopVertices3D = faceBoundLoops[0];
+						addTriangleToPolyhedronData(faceBoundLoops[0], polygon, polygonIndices);
+					}
+					else
+					{
+						// general case: arbitrary number of vertices, possible inner bound (= hole)
+						//  -> elaborate triangulation with respect to arbitrary number of vertices and holes is necessary;
 
-					// Save polygon indices of merged vertices
-					std::map<uint32_t, uint32_t> mergedIndices;
-					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> modBounds;
-					modBounds.reserve(2);
-					bool faceLoopReversed = false;
+						addArbitraryFaceToPolyhedronData(face, faceBoundLoops, polygon, polygonIndices);
+					}
+				}
+				
+				void convertIfcFaceBoundList(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
+					const carve::math::Matrix& pos,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops) const noexcept(true)

```

Comment:
Perhaps the `faceBoundLoops` can be made a return parameter?

Reply:
Done in commit d9303cc.